test/library/characterization/test_multi_state_discrimination.py

```diff
@@ -0,0 +1,77 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Test the multi state discrimination experiments."""
+from test.base import QiskitExperimentsTestCase
+from ddt import ddt, data
+
+from qiskit import pulse
+from qiskit_experiments.calibration_management import Calibrations
+from qiskit_experiments.library import MultiStateDiscrimination
+from qiskit_experiments.test.pulse_backend import SingleTransmonTestBackend
+
+
+@ddt
+class TestMultiStateDiscrimination(QiskitExperimentsTestCase):
+    """Tests of the multi state discrimination experiment."""
+
+    def setUp(self):
+        """Setup test variables."""
+        super().setUp()
+
+        self.backend = SingleTransmonTestBackend(noise=False)
+
+        # Build x12 schedule
+        self.qubit = 0
+        cals = Calibrations.from_backend(self.backend)
+
+        anharm = self.backend.anharmonicity
+
+        d0 = pulse.DriveChannel(self.qubit)
+
+        sch_map = self.backend.defaults().instruction_schedule_map
+        pulse_x = sch_map.get('x', (self.qubit,)).instructions[0][1].pulse
+        amp_x = pulse_x.amp
+        dur_x = pulse_x.duration
+        sigma_x = pulse_x.sigma
+        beta_x = pulse_x.beta
+        with pulse.build(name='x12') as x12:
+            pulse.shift_frequency(anharm, d0)
+            pulse.play(pulse.Gaussian(dur_x, amp_x * self.backend.rabi_rate_12, sigma_x, beta_x),
+                       d0)
+            pulse.shift_frequency(-anharm, d0)
+
+        cals.add_schedule(x12, qubits=self.qubit, num_qubits=1)
+
+        self.schedules = {'x12': cals.get_schedule('x12', self.qubit)}
+
+    @data(2, 3)
+    def test_circuit_generation(self, n_states):
+        """Test the experiment circuit generation"""
+        exp = MultiStateDiscrimination(self.qubit, n_states=n_states, backend=self.backend,
+                                       schedules=self.schedules)
+        self.assertEqual(len(exp.circuits()), n_states)
+
+        # check the metadata
+        self.assertEqual(exp.circuits()[-1].metadata['label'], n_states - 1)
+
+    @data(2, 3)
+    def test_discrimination_analysis(self, n_states):
+        """Test the discrimination analysis"""
+        exp = MultiStateDiscrimination(self.qubit, n_states=n_states, backend=self.backend,
+                                       schedules=self.schedules)
+
+        exp_data = exp.run()
+
+        fidelity = exp_data.analysis_results('fidelity').value
+
+        self.assertGreaterEqual(fidelity, 0.96)
```

Comment:
Can we add an assert somewhere that this discriminated `n_states`? Is that possible?

Reply:
Added in [833288e](https://github.com/Qiskit/qiskit-experiments/pull/936/commits/833288e30f24f89e58fc42c65956ba0532f7643f).